#7869 Snapping interaction implementation

alexander-fedorenko opened the pull request:
## Description
Snapping interaction implementation.

- Togglable snapping interaction.
- By default when activated it snaps to the currently selected layer + to the drawn geometry.
- Tool has dropdown listing available snapping layers (scrollable when list is too long) and snapping tool options.
- Vector-based layers are using existing source of the layers already rendered on the map. WMS layers using custom source with "bbox" loading strategy by default. It will make layer load layer features that are inside current bounding box. Snapping interaction and layer data are updated on viewport change.
- Snapping interaction configuration can be set via localConfig.json and overwritten via tool dropdown.
- Tool will display that snapping data is being loaded while vector data is loading.

localConfig.json example

```js
          {
            "name": "FeatureEditor",
            "cfg": {
              "snapTool": true,
              "snapConfig": {
                "vertex": true,
                "edge": true,
                "pixelTolerance": 10,
                "additionalLayers": [
                  "__URBANISME_RASTER_LAYER__"
                ],
                "strategy": "bbox"
              }
            }
          },
```
- Supported loading strategies for WMS layers: bbox, all
- "additionalLayers" array could be used if snapping should be available for layers that are not available in TOC. For instance, Urbanisme app. After being added into list layer will also appear in snap tool configuration dropdown.

**Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines: https://github.com/geosolutions-it/MapStore2/blob/master/CONTRIBUTING.md
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


**What kind of change does this PR introduce?** (check one with "x", remove the others)
 - [ ] Bugfix
 - [x] Feature
 - [ ] Code style update (formatting, local variables)
 - [ ] Refactoring (no functional changes, no api changes)
 - [ ] Build related changes
 - [ ] CI related changes
 - [ ] Other... Please describe:

<!-- add here the ReadTheDocs link (if needed) -->

## Issue

**What is the current behavior?**
<!-- You can also link to an existing issue here -->
#7869

**What is the new behavior?**
As per description above

## Breaking change
**Does this PR introduce a breaking change?** (check one with "x", remove the other)
 - [ ] Yes, and I documented them in migration notes
 - [x] No

<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications -->

## Other useful information

offtherailz commented on web/client/components/data/featuregrid/toolbars/Toolbar.jsx:
```diff
@@ -163,7 +166,73 @@ const standardButtons = {
+            availableSnappingLayers?.length > 1 ?
+                (
+                    <>
+                        <label className="control-label"><Message msgId="featuregrid.toolbar.snapToLayer"/></label>
+                        <div className="layers" style={{"maxHeight": pluginCfg?.snapConfig?.layersListHeight ?? '130px', 'overflowY': 'auto'}}>
+                            { availableSnappingLayers.map(option => <MenuItem active={option.active} onClick={() => !option.active && events.setSnappingLayer(option.value)}>{option.label}</MenuItem>) }
+                        </div>
```
Here use a simple "Select" component, so you don't have to fix the height or handle it. 
By default the current layer must be selected

alexander-fedorenko replied:
Replaced list of items with react-select. As I mentioned previously - it's impossible to render v1 in a portal, so we should think of upgrading this dependency at some point.
Right now dropdown maximum height is capped by feature editor height and react-select options list is capped by the height of dropdown. Scrollbar will appear if either of two cannot fit in feature editor / in dropdown.

offtherailz commented on web/client/plugins/FeatureEditor.jsx:
```diff
@@ -111,7 +111,17 @@ const Dock = connect(createSelector(
   *     },
-  *   "editingAllowedRoles": ["ADMIN"]
+  *   "editingAllowedRoles": ["ADMIN"],
+  *   "snapTool": true,
```
This should be true by default. Please document in props.

alexander-fedorenko replied:
So snap tool should be available by default and invisible only if explicitly set to false in localConfig? I've updated code to make it work this way, just asking to be 100% sure that's the case.

offtherailz replied:
@tdipisa snappint tool should be always available, also instandard mapstore, isn't it? 
I think it makes easier to test.

offtherailz commented on web/client/plugins/FeatureEditor.jsx:
```diff
@@ -111,7 +111,17 @@ const Dock = connect(createSelector(
+  *     "vertex": true,
+  *     "edge": true,
+  *     "pixelTolerance": 10,
+ *      "additionalLayers": [
+ *         "ADDITIONAL_LAYER_ID"
+ *      ],
```
is this still necessary?

alexander-fedorenko replied:
Yes, this one can be useful if someone needs to snap to the layer that is not available in TOC. Like urbanisme layer with recent changes. This is configurable only in localConfig and will bring layer into snapping options list if additional layer is active & supported type (vector / wms + wfs)

offtherailz commented on web/client/selectors/draw.js:
```diff
@@ -5,10 +5,60 @@
+    return [{value: id, label: 'Current layer', active: id === snappingId}].concat(
+        layersList.map((layer) =>
+            layer.id !== id
+                && ['wms', 'wfs', 'vector'].includes(layer?.type)
+                && layer.group !== 'background'
+                && layer.visibility ? {
+                    value: layer.id,
+                    label: layer.title ?? layer.name,
+                    active: layer.id === snappingId
+                } : false).filter(Boolean)
```
- [ ] WMS layers are valid ONLY if they have the `search` url defined as "wfs".

alexander-fedorenko replied:
Corrected condition for WMS layers

offtherailz commented on web/client/translations/data.it-IT.json:
```diff
@@ -1678,7 +1678,16 @@
                 "disableTimeSync": "Non sincronizzare con l'istante o l'intervallo temporale selezionato",
-                "createNewChart": "crea un grafico per il livello selezionato"
+                "createNewChart": "crea un grafico per il livello selezionato",
+                "enableSnapping": "Attivare lo scatto",
+                "disableSnapping": "Disattivare lo snap",
+                "snapToLayer": "Aggancia al livello",
+                "snappingSettings": {
+                    "header": "Impostazioni",
+                    "vertex": "Vertice",
+                    "edge": "Bordo",
+                    "pixelTolerance": "Tolleranza pixel"
+                  }
```
- [ ] This is for italian.
```suggestion
                "enableSnapping": "Attiva snapping",
                "disableSnapping": "Disattiva snapping",
                "snapToLayer": "Aggancia al livello",
                "snappingSettings": {
                    "header": "Impostazioni",
                    "vertex": "Vertice",
                    "edge": "Bordo",
                    "pixelTolerance": "Tolleranza (pixel)"
                  }
```

offtherailz commented on web/client/translations/data.fr-FR.json:
```diff
@@ -1679,7 +1679,16 @@
                 "disableTimeSync": "Ne pas synchroniser avec l'heure ou l'intervalle sélectionné",
-                "createNewChart": "Créer un graphique pour le calque sélectionné"
+                "createNewChart": "Créer un graphique pour le calque sélectionné",
+                "enableSnapping": "Activer l'accrochage",
+                "disableSnapping": "Désactiver la capture",
```
- [ ] these doesn't seem constistent.  Trying to guess :smile: 
```suggestion
                "enableSnapping": "Activer l'accrochage",
                "disableSnapping": "Désactiver l'accrochage",
```

offtherailz commented on web/client/translations/data.fr-FR.json:
```diff
@@ -1679,7 +1679,16 @@
+                  "header": "Paramètres",
+                  "vertex": "Vertex",
+                  "edge": "Bord",
+                  "pixelTolerance": "Tolérance de pixels"
```
- [ ] Same for all translations

```suggestion
                  "pixelTolerance": "Tolérance (pixel)"
```

offtherailz reviewed (changes requested):
I love it :heart:  It is very cool 

 See my comments inline. Moreover the following ponts (Sorry if some of them are duplicated, github is not working properly today): 
- [ ] When I disable snapping, the editing resets, this is annoying, can we avoid it ?
https://user-images.githubusercontent.com/1279510/158828687-eabdd672-83fd-4c34-abb6-2f2c0d2ccc6a.mp4
- [ ] We should also set a limit of feature to download, configurable at least from plugin config, or from UI. 
- [ ] Use a select for layer, instead of menu items, for the layers. This could help to remove all the custom .less and the ">1" check (always show the select, even with 1 element only). 
- [ ] should use the wfsURL, if present. Actually we use "search" entry of type . This is not well understandable but it is the current implementation. Instead you are using the layer URL and this makes the request to "wms" path. It is ok for geoserver, not always for other layers.  

For the moment I think this is enough. 

For the future (not sure we planned or not): 
- [ ] Can we select the strategy from the UI ? By default the one configured ("bbox" if not configured).

alexander-fedorenko commented on web/client/components/map/openlayers/DrawSupport.jsx:
```diff
@@ -381,12 +421,92 @@ export default class DrawSupport extends React.Component {
+                                }
+                            };
+                            isLoading();
+                            xhr.send();
```
Thanks, I've changed it to axios. Are you sure about the version from libs? axios was there in imports even before my changes and version is the one from node_modules.

alexander-fedorenko commented on web/client/components/map/openlayers/DrawSupport.jsx:
```diff
@@ -381,12 +421,92 @@ export default class DrawSupport extends React.Component {
+                        format: new GeoJSON(),
+                        loader: function(extent, resolution, projection) {
+                            const proj = projection.getCode();
+                            const url = getFeatureURL(getLayerUrl(snappingLayerInstance), snappingLayerInstance.name, {
+                                version: '1.1.0',
+                                outputFormat: 'application/json',
+                                srsname: proj,
+                                bbox: extent.join(',') + ',' + proj
+                            });
```
Sorry, I didn't understand this point. Is this the wfsURL in store "state.query.wfsURL"?

alexander-fedorenko commented on web/client/plugins/map/index.js:
```diff
@@ -91,14 +96,22 @@ const pluginsCreator = (mapType, actions) => {
+                snappingLayerInstance: snappingLayerSelector(state),
+                snappingLayerType: snappingLayerType(state),
+                snappingLayerData: snappingLayerDataSelector(state)
+            }) || {}, {
```
Thanks, corrected it.

alexander-fedorenko commented on web/client/reducers/draw.js:
```diff
@@ -43,6 +50,39 @@ function draw(state = initialState, action) {
+        return {
+            ...state,
+            snappingLayer: action.snappingLayer,
+            snappingShouldRefresh: true,
```
Yes, this is leftover from previous implementation, Cleaned up this one and all other leftovers.

alexander-fedorenko commented on web/client/components/map/openlayers/DrawSupport.jsx:
```diff
@@ -102,7 +121,10 @@ export default class DrawSupport extends React.Component {
-        onEndDrawing: () => {}
+        onEndDrawing: () => {},
+        onUpdateSnappingLayer: () => {},
+        setSnappingShouldRefresh: () => {},
```
Cleaned up

alexander-fedorenko commented on web/client/components/map/openlayers/DrawSupport.jsx:
```diff
@@ -85,7 +89,22 @@ export default class DrawSupport extends React.Component {
+        onUpdateSnappingLayer: PropTypes.func,
+        onSnappingRequestWMSFeatures: PropTypes.func,
+        onRefreshSnappingLayer: PropTypes.func,
+        setSnappingShouldRefresh: PropTypes.func,
```
Cleaned up

alexander-fedorenko commented on web/client/components/map/openlayers/DrawSupport.jsx:
```diff
@@ -85,7 +89,22 @@ export default class DrawSupport extends React.Component {
+        snappingLayer: PropTypes.object,
+        snappingLayerData: PropTypes.object,
+        isSnappingLoading: PropTypes.bool,
+        snappingShouldRefresh: PropTypes.bool,
```
Cleaned up

alexander-fedorenko commented on web/client/components/map/openlayers/DrawSupport.jsx:
```diff
@@ -381,12 +421,92 @@ export default class DrawSupport extends React.Component {
+        }
+    }
+
+    addSnapInteraction = (newProps) => {
```
Updated documentation and slightly modified function definition.

alexander-fedorenko commented on web/client/actions/draw.js:
```diff
@@ -74,3 +82,39 @@ export function setCurrentStyle(style) {
+        type: SET_SNAPPING_DEFAULT_CONFIG,
+        pluginCfg
+    };
+}
```
Thanks, simplified, now only `setSnappingConfig` is in use

alexander-fedorenko commented on web/client/components/data/featuregrid/toolbars/Toolbar.jsx:
```diff
@@ -163,7 +166,73 @@ const standardButtons = {
+        pullLeft
+    >
+        {
+            availableSnappingLayers?.length > 1 ?
```
Replaced it with unconditional rendering of select

alexander-fedorenko commented:
Amended old and added new tests.
Removed annoying behavior of draw session reset when span tool is toggled. It won't re-initialize snap interaction if it's not the case of layer change or configuration change. Simple toggle with the same layer will make interaction disabled/enabled back.

Open questions:
- axios on DrawSupport, do we need to switch to the version from libs?
- defaults for snap tool availability. Should it be visible unless disabled in localConfig?
- not very clear if wfsURL is the one that is in store. It is indeed used in query observables, didn't have a chance to find how it gets to store.
- was about to ask about limit of features but just found that it can be easily changed by setting "maxFeatures" in WFS query.

offtherailz reviewed (changes requested):
Hi. I pushed some changes to clean up and fix minor issues. 

Maybe we could use a simple select to avoid issues in the dropdown menu. Anyway now I reduced a lot the style necessary to it. 


- [ ] I'd like to merge, but I only noticed that when you toggle the snap the editing is not reset anymore, but when I toggle the other options (e.g. change layer, toggle settings) the editing is reset. Is it possible to avoid it. The rest is perfect.

offtherailz reviewed (changes requested):
It looks great. I only found a bug while testing here and there.
Sometimes the layer title can be an object (localized). like in this [map](http://localhost:8081/#/viewer/openlayers/24794).
In this case (selecting the second layer and trying to edit) the application crashes.
In this case you should do something like 
```
const realTitle = isObject(node.title) ? node.title[currentLocale] || node.title.default : node.title;
```

The rest seems very good

http://localhost:8081/#/viewer/openlayers/24794

alexander-fedorenko commented:
Thanks! Applied minor change, title in layers list should be locale-sensitive.
@offtherailz

offtherailz reviewed (changes requested):
We are almost there. I clean up a little the code for better understanding, centralizing the management of the snap interaction and removed the self snapping that caused issues (and is not required for the moment, so let's ignore it

Here the things:
- [x] When deactivate snap - change layer - reactivate snap the layer is not updated --> i fixed it with the refactor I committed
- [x] When changing feature of editing after a first snap activation the application crashes (because the drawSource raised some errors) I removed the self snapping on drawSource and this solved the bug
- [x]  If you start editing with snapping, then close the feature grid and delete the layer, the application crashes. I'd like you to check it. In general we should allow a more transparent clean up. I tried adding removeSnapInteraction on clean but I see this is not enough
- [ ] Snap  feature download starts after first change, so you can't start snapping from the beginning

alexander-fedorenko commented:
Remaining items should be resolved.
- Added extra checks for map layers existence prior to attempt to instantiate snap interaction.
- Snapping forcefully toggled off when featuregrid is closed; Snapping layer is reset to initial state when featuregrid is closed.

alexander-fedorenko commented:
@offtherailz please review. Merging to the master should not lead to a problem as readthedocs build is fixed on master with #8025

tdipisa commented:
@ElenaGallo please test this in DEV and let us know for the backport

ElenaGallo commented:
Hi @alexander-fedorenko, some notes after my test:

1_ Trying to change the Tolerance the popup closes, either by clicking in the box or after clicking on the arrows.

https://user-images.githubusercontent.com/56537133/160596189-aed039ea-2a31-4c6b-8cb2-298497e0bea4.mov

2_ The bottom popup border should be on all sides of the popup

![2](https://user-images.githubusercontent.com/56537133/160596204-18cde8c7-4e6d-4d0f-b931-c09efc1d0ac8.png)

3_ The tooltip for the Snap button is missing

4_ When the Snaps button is disabled, the snap lines remain highlighted on the map

https://user-images.githubusercontent.com/56537133/160596344-f865ab4b-fbb8-4140-9941-e92eecf89878.mov

5_ Closing the edit mode the snap button is not disabled and entering edit mode again the button is already enabled

https://user-images.githubusercontent.com/56537133/160596400-757b4e07-cf5b-4919-bb23-b572568b044c.mov
